unixfs/mod/dagmodifier.go

```diff
@@ -0,0 +1,464 @@
+		}
+
+		dm.read = dr
+	}
+
+	n, err := dm.read.Read(b)
+	dm.curWrOff += uint64(n)
+	return n, err
+}
+
+// splitBytes uses a splitterFunc to turn a large array of bytes
+// into many smaller arrays of bytes
+func (dm *DagModifier) splitBytes(in io.Reader) ([]u.Key, error) {
+	var out []u.Key
+	blks := dm.splitter.Split(in)
+	for blk := range blks {
+		nd := help.NewUnixfsNode()
+		nd.SetData(blk)
+		dagnd, err := nd.GetDagNode()
+		if err != nil {
+			return nil, err
+		}
+
+		k, err := dm.dagserv.Add(dagnd)
+		if err != nil {
+			return nil, err
+		}
+		out = append(out, k)
+	}
+	return out, nil
+}
+
+// GetNode gets the modified DAG Node
+func (dm *DagModifier) GetNode() (*mdag.Node, error) {
+	err := dm.Flush()
+	if err != nil {
+		return nil, err
+	}
+	return dm.curNode.Copy(), nil
+}
+
+// HasChanges returned whether or not there are unflushed changes to this dag
+func (dm *DagModifier) HasChanges() bool {
+	return dm.wrBuf != nil
+}
+
+func (dm *DagModifier) Seek(offset int64, whence int) (int64, error) {
+	err := dm.Flush()
+	if err != nil {
+		return 0, err
+	}
+
+	switch whence {
+	case os.SEEK_CUR:
+		dm.curWrOff += uint64(offset)
+		dm.writeStart = dm.curWrOff
+	case os.SEEK_SET:
+		dm.curWrOff = uint64(offset)
+		dm.writeStart = uint64(offset)
+	case os.SEEK_END:
+		return 0, errors.New("SEEK_END currently not implemented")
+	default:
+		return 0, errors.New("unrecognized whence")
+	}
+
+	if dm.read != nil {
+		_, err = dm.read.Seek(offset, whence)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	return int64(dm.curWrOff), nil
+}
+
+func (dm *DagModifier) Truncate(size int64) error {
+	err := dm.Flush()
+	if err != nil {
+		return err
+	}
+
+	realSize, err := dm.Size()
+	if err != nil {
+		return err
+	}
+
+	// Truncate can also be used to expand the file
+	if size > int64(realSize) {
+		return dm.expandSparse(int64(size) - realSize)
+	}
+
+	nnode, err := dagTruncate(dm.curNode, uint64(size), dm.dagserv)
+	if err != nil {
+		return err
+	}
+
+	_, err = dm.dagserv.Add(nnode)
+	if err != nil {
+		return err
+	}
+
+	dm.curNode = nnode
+	return nil
+}
+
+// dagTruncate truncates the given node to 'size' and returns the modified Node
+func dagTruncate(nd *mdag.Node, size uint64, ds mdag.DAGService) (*mdag.Node, error) {
+	if len(nd.Links) == 0 {
+		// TODO: this can likely be done without marshaling and remarshaling
+		pbn, err := ft.FromBytes(nd.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		nd.Data = ft.WrapData(pbn.Data[:size])
+		return nd, nil
+	}
+
+	var cur uint64
+	end := 0
+	var modified *mdag.Node
+	ndata := new(ft.FSNode)
+	for i, lnk := range nd.Links {
+		child, err := lnk.GetNode(ds)
+		if err != nil {
+			return nil, err
+		}
+
+		childsize, err := ft.DataSize(child.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		if size < cur+childsize {
+			nchild, err := dagTruncate(child, size-cur, ds)
+			if err != nil {
+				return nil, err
+			}
+
+			// TODO: sanity check size of truncated block
```

Comment:
problem is its not free to do so, and ive already verified that it is correct in all test cases. I should probably just remove this TODO